[RFC][Routing] Add warnings to "debug:router" command.

SpacePossum opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no 
| Deprecations? | no
| Tests pass?   | we'll see ^_^
| License       | MIT
| Doc PR        | no update needed

When using `debug:router`; this adds colors to warn a user about possible typo's in the routing setup.
This is only done on the text format as to not break any BC.

I thought a picture might be more useful than a full description:

![debug_router](https://user-images.githubusercontent.com/10462973/41347381-c05ed77e-6f09-11e8-8670-13f31cd474e7.png)

SpacePossum committed:
Add warnings to "debug:router" command.

stof commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -32,6 +35,22 @@
+
+    public function __construct()
+    {
+        $this->validator = Validation::createValidatorBuilder()->getValidator();
```
what if the validator component is not installed ?

stof commented on src/Symfony/Bundle/FrameworkBundle/composer.json:
```diff
@@ -27,7 +27,8 @@
         "symfony/finder": "~3.4|~4.0",
-        "symfony/routing": "^4.1"
+        "symfony/routing": "^4.1",
+        "symfony/validator": "^4.1"
```
-1 for making it a hard requirement. This defeats the work done in 3.4 to reduce the number of mandatory deps in a Symfony project (and defeats the auto-enabling of the component as it would now always be installed)

stof replied:
you might want to duplicate the regex in the descriptor instead to avoid a strong dependency (and maybe add a unit test ensuring that the regex stays in sync, but this might be a pain if different versions get installed for FrameworkBundle and the component)

iltar replied:
What about an interface where you can have an adapter for to validate using the validation component instead? Making the "url is invalid" feature automatically enable when the validation component exists, and do nothing when it isn't (perhaps with a suggestion to install it)

stof replied:
well, if you only want to disable the feature, there is no need for an adapter. You would only need to make an existence check in the constructor, and skip the validation if you have `null` as validator.

iltar replied:
It's nice for other frameworks if they wish to provide an integration without the Symfony validator

stof replied:
this descriptor is in FrameworkBundle. That's the package which is not meant to be reused by other frameworks, as it defines the Symfony framework experience.

iltar commented:
This looks like a really fancy feature, thanks!

SpacePossum committed:
revert moving "symfony/validator" from "require-dev" to "require".

SpacePossum committed:
CS - use long array notation

SpacePossum committed:
TMP - add details about failing test, might not be related.

SpacePossum committed:
validate paths without depending on "symfony/validator".

SpacePossum commented:
Thanks for the review and feedback :)
I've updated the code to no longer rely on "symfony/validator". After puzzling things around I've to say I'm not that happy with copy-pasting the pattern to validate with, however it might be the best solution there is now.

Updated sample for clarity:

![debug_router2](https://user-images.githubusercontent.com/10462973/41398410-87dd72e4-6fb7-11e8-9e51-72fbc746973e.png)

SpacePossum committed:
Yoda :)

thats what you get for writing a Yoda fixer ;)

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,75 @@ private function writeText(string $content, array $options = array())
+
+    private function describeMethods(array $methods)
+    {
+        if (0 === count($methods)) {
```
`if (!$methods) {`

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,75 @@ private function writeText(string $content, array $options = array())
+            $methodDescription .= isset($knownMethods[$method])
+                ? $method
+                : '<error>'.$method.'</error>'
+            ;
```
full  expression should be on one line

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,75 @@ private function writeText(string $content, array $options = array())
+            ((?:\# (?:[\pL\pN\-._\~!$&\'()*+,;=:@/?]|%%[0-9A-Fa-f]{2})* )?) # a fragment (optional)
+        $~ixu';
+
+        if (1 !== preg_match($pattern, preg_replace('#\{\w+\}#', 'X', $path), $matches)) {
```
`if (!preg_match(...`

nicolas-grekas commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,75 @@ private function writeText(string $content, array $options = array())
+
+        // while a valid path it is still likely a typo
+        $doubleSlashIndex = strpos($path, '//');
+        if (false !== $doubleSlashIndex) {
```
`if (false !== $doubleSlashIndex = strpos($path, '//')) {`

nicolas-grekas commented on src/Symfony/Component/Config/Tests/Resource/DirectoryResourceTest.php:
```diff
@@ -22,9 +22,14 @@ protected function setUp()
         $this->directory = sys_get_temp_dir().DIRECTORY_SEPARATOR.'symfonyDirectoryIterator';
         if (!file_exists($this->directory)) {
-            mkdir($this->directory);
+            if (false === mkdir($this->directory)) {
```
`if (!mkdir(...`

nicolas-grekas commented on src/Symfony/Component/Config/Tests/Resource/DirectoryResourceTest.php:
```diff
@@ -22,9 +22,14 @@ protected function setUp()
+            }
+        }
+
+        if (false === touch($this->directory.'/tmp.xml')) {
```
`if (!touch(...`

nicolas-grekas reviewed (approved):
LGTM thank you.
(with minor CS comments)

SpacePossum committed:
update CS

chalasr commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,70 @@ private function writeText(string $content, array $options = array())
         );
     }
+
+    private function describeMethods(array $methods)
```
The descriptor does not only deal with routing (DI, event dispatching), I would rename to `describeRouteMethods()`/`describeRoutePath()`

SpacePossum committed:
rename methods to `describeRoute[X]()`

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,70 @@ private function writeText(string $content, array $options = array())
+    private function describeRoutePath(string $path)
+    {
+        if ('' === $path || '/' === $path) {
+            return $path; // cheap short cuts
```
We can remove the `cheap short cuts` comment because it's not needed to understand the code.

javiereguiluz commented on src/Symfony/Bundle/FrameworkBundle/Console/Descriptor/TextDescriptor.php:
```diff
@@ -470,4 +470,70 @@ private function writeText(string $content, array $options = array())
+            return '<info>ANY</info>';
+        }
+
+        $knownMethods = array(
```
We could use `Request::METHOD_GET`, `Request::METHOD_POST`, etc. By the way, Symfony's Request also supports `PURGE` method.

javiereguiluz reviewed (changes requested):
This can be useful to spot some errors. I'm not sure about its discoverability. I see some parts of `debug:router` in yellow ... but the reason is not explained anywhere.

Maybe we could add instead a `lint:routes` command?

SpacePossum committed:
remove comment and 'PURGE'

SpacePossum commented:
>Maybe we could add instead a lint:routes command?

If others feel that would be a better fit as well than I can try to write the command.

stof commented:
> If others feel that would be a better fit as well than I can try to write the command.

That could make sense indeed, as it would mean we can provide more details about the error instead of only coloring it (which is lost in terminal without color support, or for color-blind people, and is not clear for people not reading the feature announcement), and we can make the exit code a failure one in case of mistakes (making this feature useful for CI purposes too)

SpacePossum commented:
thanks all, when I find the time I try to make the `route:lint` thing :+1:
